src/tactic/ring_exp.lean

```diff
@@ -0,0 +1,1419 @@
+  reader_t.lift $ state_t.put atoms',
+  pure atm
+
+/--
+  Treat the expression atomically: as a coefficient or atom.
+
+  Handles cases where `eval` cannot treat the expression as a known operation
+  because it is just a number or single variable.
+-/
+meta def eval_base (ps : expr) : ring_exp_m (ex sum) :=
+match ps.to_rat with
+| some ⟨0, 1, _, _⟩ := ex_zero
+| some x := ex_coeff x >>= prod_to_sum
+| none := do
+  a ← resolve_atom ps,
+  atom_to_sum a
+end
+
+lemma negate_pf {α} {ps ps' : α} [comm_ring α] : (-1) * ps = ps' → -ps = ps' := by simp
+/--
+  Negate an expression by multiplying with `-1`.
+
+  Only works if there is a `comm_ring` instance; otherwise it will `fail`.
+-/
+meta def negate (ps : ex sum) : ring_exp_m (ex sum) := do
+  ctx ← get_context,
+  match ctx.info_b.cr_instance with
+  | none := lift $ fail "internal error: negate called in semiring"
+  | (some cr_instance) := do
+    minus_one ← ex_coeff (-1) >>= prod_to_sum,
+    ps' ← mul minus_one ps,
+    ps_pf ← ps'.proof_term,
+    -- We can't use mk_app at the next line because it would cause a timeout.
+    pf ← lift $ to_expr ``(@negate_pf _ _ _ %%cr_instance %%ps_pf),
+    ps'_o ← lift $ mk_app ``has_neg.neg [ps.orig],
+    pure $ ps'.set_info ps'_o pf
+  end
+
+lemma inverse_pf [division_ring α] {ps ps_u ps_p e' e'' : α} :
+  ps = ps_u → ps_u = ps_p → ps_p ⁻¹ = e' → e' = e'' → ps ⁻¹ = e'' :=
+by cc
+/--
+  Invert an expression by simplifying, applying `has_inv.inv` and treating the result as an atom.
+
+  Only works if there is a `division_ring` instance; otherwise it will `fail`.
+-/
+meta def inverse (ps : ex sum) : ring_exp_m (ex sum) := do
+  ctx ← get_context,
+  dri ← match ctx.info_b.dr_instance with
+  | none := lift $ fail "division is only supported in a division ring"
+  | (some dri) := pure dri
+  end,
+  (ps_simple, ps_simple_pf) ← ps.simple,
+  e ← lift $ mk_app ``has_inv.inv [ps_simple],
+  (e', e_pf) ← lift (norm_num.derive e) <|> ((λ e_pf, (e, e_pf)) <$> lift (mk_eq_refl e)),
+  e'' ← eval_base e',
+  ps_pf ← ps.proof_term,
+  pf ← mk_proof_or_refl e''.pretty ``inverse_pf
+    [dri, ps.orig, ps.pretty, ps_simple, e', e''.pretty, ps_pf, ps_simple_pf, e_pf]
+    [e''.info],
+  e''_o ← lift $ mk_app ``has_inv.inv [ps.orig],
+  pure $ e''.set_info e''_o pf
+
+lemma sub_pf [comm_ring α] {ps qs psqs : α} : ps + -qs = psqs → ps - qs = psqs := id
+lemma div_pf [division_ring α] {ps qs psqs : α} : ps * qs⁻¹ = psqs → ps / qs = psqs := id
+
+end operations
+
+section wiring
+/-!
+  ## `wiring` section
+  This section deals with going from `expr` to `ex` and back.
+
+  The main attraction is `eval`, which uses `add`, `mul`, etc.
+  to calculate an `ex` from a given `expr`.
+  Other functions use `ex`es to produce `expr`s together with a proof,
+  or produce the context to run `ring_exp_m` from an `expr`.
+-/
+
+open tactic
+open ex_type
+
+meta def eval : expr → ring_exp_m (ex sum)
+| e@`(%%ps + %%qs) := do
+  ps' ← eval ps,
```

Comment:
These lines can be written more concisely as `add <$> ps' <*> qs'` if monad glyphs are your thing. Not a necessary change though, it's a matter of taste. Same applies below.

Reply:
Unfortunately, the type of `add` is `ex sum → ex sum → ring_exp_m (ex sum)`, so we really need a monadic bind here.

Reply:
In commit f9b72160, I switched other terms to use applicative notation (where I could do so without introducing a λ term). I'm unsure which version I prefer if the other code is already in `do` notation, so I would be happy to revert that commit.

Reply:
Ah, I see. It's your call. There are no style guidelines about this and you'll see both in the library.